File: WordPress/Classes/Services/WordPressComOAuthClientFacade.m
#import "WordPressComOAuthClientFacade.h"
#import "WordPress-Swift.h"

@interface WordPressComOAuthClientFacade(Internal)

@property (nonatomic, strong) WordPressComOAuthClient *client;

@end

@implementation WordPressComOAuthClientFacade

- (instancetype)initWithClient:(NSString *)client secret:(NSString *)secret
{
    NSParameterAssert(client);
    NSParameterAssert(secret);
    self = [super init];
    if (self) {
        self.client = [WordPressComOAuthClient clientWithClientID:client secret:secret];
    }

    return self;
}

- (instancetype)init {
    NSAssert(false, @"Please initializer WordPressComOAuthClientFacade with the ClientID and Secret!");
    return nil;
}

- (void)authenticateWithUsername:(NSString *)username
                        password:(NSString *)password
                 multifactorCode:(NSString *)multifactorCode
                         success:(void (^)(NSString *authToken))success
                needsMultiFactor:(void (^)(void))needsMultifactor
                         failure:(void (^)(NSError *error))failure
{
    [self.client authenticateWithUsername:username password:password multifactorCode:multifactorCode success:success failure:^(NSError *error) {
        if (error.code == WordPressComOAuthErrorNeedsMultifactorCode) {
            if (needsMultifactor != nil) {
                needsMultifactor();
            }
        } else {
            if (failure != nil) {
                failure(error);
            }
        }
    }];
}

- (void)requestOneTimeCodeWithUsername:(NSString *)username
                              password:(NSString *)password
                               success:(void (^)(void))success
                               failure:(void (^)(NSError *error))failure
{
    [self.client requestOneTimeCodeWithUsername:username password:password success:success failure:failure];
}

- (void)requestSocial2FACodeWithUserID:(NSInteger)userID
                                nonce:(NSString *)nonce
                                 success:(void (^)(NSString *newNonce))success
                                 failure:(void (^)(NSError *error, NSString *newNonce))failure
{
    [self.client requestSocial2FACodeWithUserID:userID nonce:nonce success:success failure:failure];
}

- (void)authenticateWithGoogleIDToken:(NSString *)token
                              success:(void (^)(NSString *authToken))success
                     needsMultiFactor:(void (^)(NSInteger userID, SocialLogin2FANonceInfo *nonceInfo))needsMultifactor
          existingUserNeedsConnection:(void (^)(NSString *email))existingUserNeedsConnection
                              failure:(void (^)(NSError *error))failure
{
    [self.client authenticateWithIDToken:token success:success needsMultifactor:needsMultifactor existingUserNeedsConnection:existingUserNeedsConnection failure:failure];
}

- (void)authenticateSocialLoginUser:(NSInteger)userID
                           authType:(NSString *)authType
                        twoStepCode:(NSString *)twoStepCode
                       twoStepNonce:(NSString *)twoStepNonce
                            success:(void (^)(NSString *authToken))success
                            failure:(void (^)(NSError *error))failure
{
    [self.client authenticateSocialLoginUser:userID authType:authType twoStepCode:twoStepCode twoStepNonce:twoStepNonce success:success failure:failure];
}

@end

File: WordPressAuthenticator/Sources/Services/WordPressComOAuthClientFacade.swift
import Foundation
import WordPressKit

@objc public class WordPressComOAuthClientFacade: NSObject, WordPressComOAuthClientFacadeProtocol {

    private let client: WordPressComOAuthClient

    @objc required public init(client: String, secret: String) {
        self.client = WordPressComOAuthClient(
            clientID: client,
            secret: secret,
            wordPressComBaseURL: WordPressAuthenticator.shared.configuration.wpcomBaseURL,
            wordPressComApiBaseURL: WordPressAuthenticator.shared.configuration.wpcomAPIBaseURL
        )
    }

    public func authenticate(
        username: String,
        password: String,
        multifactorCode: String?,
        success: @escaping (_ authToken: String?) -> Void,
        needsMultifactor: @escaping ((_ userID: Int, _ nonceInfo: SocialLogin2FANonceInfo?) -> Void),
        failure: ((_ error: Error) -> Void)?
    ) {
        self.client.authenticate(username: username, password: password, multifactorCode: multifactorCode, needsMultifactor: needsMultifactor, success: success, failure: { error in
            if case let .endpointError(authenticationFailure) = error, authenticationFailure.kind == .needsMultifactorCode {
                needsMultifactor(0, nil)
            } else {
                failure?(error)
            }
        })
    }

    public func requestOneTimeCode(
        username: String,
        password: String,
        success: @escaping () -> Void,
        failure: @escaping (_ error: Error) -> Void
    ) {
        self.client.requestOneTimeCode(username: username, password: password, success: success, failure: failure)
    }

    public func requestSocial2FACode(
        userID: Int,
        nonce: String,
        success: @escaping (_ newNonce: String) -> Void,
        failure: @escaping (_ error: Error, _ newNonce: String?) -> Void
    ) {
        self.client.requestSocial2FACode(userID: userID, nonce: nonce, success: success, failure: failure)
    }

    public func authenticate(
        socialIDToken: String,
        service: String,
        success: @escaping (_ authToken: String?) -> Void,
        needsMultifactor: @escaping (_ userID: Int, _ nonceInfo: SocialLogin2FANonceInfo) -> Void,
        existingUserNeedsConnection: @escaping (_ email: String) -> Void,
        failure: @escaping (_ error: Error) -> Void
    ) {
        self.client.authenticate(
            socialIDToken: socialIDToken,
            service: service,
            success: success,
            needsMultifactor: needsMultifactor,
            existingUserNeedsConnection: existingUserNeedsConnection,
            failure: failure
        )
    }

    public func authenticate(
        socialLoginUser userID: Int,
        authType: String,
        twoStepCode: String,
        twoStepNonce: String,
        success: @escaping (_ authToken: String?) -> Void,
        failure: @escaping (_ error: Error) -> Void
    ) {
        self.client.authenticate(
            socialLoginUserID: userID,
            authType: authType,
            twoStepCode: twoStepCode,
            twoStepNonce: twoStepNonce,
            success: success,
            failure: failure
        )
    }

    public func requestWebauthnChallenge(
        userID: Int64,
        twoStepNonce: String,
        success: @escaping (_ challengeData: WebauthnChallengeInfo) -> Void,
        failure: @escaping (_ error: Error) -> Void
    ) {
        self.client.requestWebauthnChallenge(userID: userID, twoStepNonce: twoStepNonce, success: success, failure: failure)
    }

    public func authenticateWebauthnSignature(
        userID: Int64,
        twoStepNonce: String,
        credentialID: Data,
        clientDataJson: Data,
        authenticatorData: Data,
        signature: Data,
        userHandle: Data,
        success: @escaping (_ authToken: String) -> Void,
        failure: @escaping (_ error: Error) -> Void
    ) {
        self.client.authenticateWebauthnSignature(
            userID: userID,
            twoStepNonce: twoStepNonce,
            credentialID: credentialID,
            clientDataJson: clientDataJson,
            authenticatorData: authenticatorData,
            signature: signature,
            userHandle: userHandle,
            success: success,
            failure: failure
        )
    }

}
